Poseidon hash

tsunrise opened the pull request.

tsunrise committed:
matrix impl with Tom

tsunrise committed:
mds matrix with Tom

tsunrise committed:
milestone: halfway

tsunrise committed:
poseidon (correct) native

tsunrise committed:
nit

stechu commented on Cargo.toml:
```diff
@@ -16,17 +16,36 @@ exclude = [
+rand = "0.8"
+thiserror = "1.0.30"
+
+ark-plonk = { version = "^0.8", git = "https://github.com/rust-zkp/ark-plonk", default-features = false }
```
Is it possible to update to the new repo? https://github.com/ZK-Garage/plonk

tsunrise replied:
yes, updated

tsunrise committed:
start working on constraints

bhgomes commented on src/poseidon/matrix.rs:
```diff
@@ -0,0 +1,675 @@
+use ark_ff::PrimeField;
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct Matrix<T: Clone>(pub Vec<Vec<T>>);
```
I would suggest using a `Vec<T>` in row-major order for better performance.

tsunrise replied:
make sense (for better cache performance)

bhgomes commented on src/poseidon/matrix.rs:
```diff
@@ -0,0 +1,675 @@
+    }
+
+    pub fn is_identity(&self) -> bool {
+        if !self.is_square() {
```
Aren't all these matrices square? The transpose method for example seems to assume so. Do we need non-square matrices anywhere?

tsunrise replied:
yes all MDS matrices should be square

bhgomes replied:
Then, we should notate this and make sure that we are using this fact everywhere in the interface and the implementation.

bhgomes commented on src/poseidon/poseidon_ref.rs:
```diff
@@ -0,0 +1,218 @@
+pub struct PoseidonConstants<F: PrimeField> {
+    pub mds_matrices: MdsMatrices<F>,
+    pub round_constants: Vec<F>,
+    pub domain_tag: F,
```
What is the `domain_tag` used for?

BoyuanFeng replied:
It is concatenated with hash input as the final hash input.

For example, if arity is 2, we have 2+1 hash inputs.

tsunrise replied:
for now, domain tag is just `2^ARITY - 1`

bhgomes replied:
Is this in the original spec? I thought you were supposed to zero-pad the input like this:

```
fn hash(x: T, y: T) -> T {
    poseidon_permutation_of_width_3(0, x, y)
}
```

bhgomes reviewed (commented):
Just some superficial things I wanted to ask about right now. Didn't finish reviewing every detail.

tsunrise committed:
spec on plonk gate

tsunrise committed:
api changes

tsunrise committed:
nit

tsunrise committed:
nit

tsunrise committed:
circuits for poseidon_ref

tsunrise committed:
some specialization for plonk

tsunrise committed:
add tests for circuits of poseidon_ref

tsunrise committed:
add tests for circuits of poseidon_ref

tsunrise committed:
poseidon_ref

tsunrise committed:
poseidon_ref

tsunrise committed:
poseidon_ref

tsunrise committed:
sparse mat halfway

Commit:
implement preprocessing for round constants

tsunrise committed:
skeleton for optimized poseidon

tsunrise committed:
poseidon optimized (has some bugs)

tsunrise committed:
field conversion from ff to ark_ff

Commit:
convert neptune constants

Commit:
convert constants

tsunrise committed:
add unit tests to compare with neptune

tsunrise committed:
plonk_ref is now correct

Commit:
fix a bug in optimized poseidon

Commit:
add plonk & r1cs impl for poseidon hash; add tests

tsunrise committed:
bugfix

tsunrise committed:
move "output_hash" to spec

because in plonk optimizations, we might need to combine some ARC with MDS for arity 3.

tsunrise committed:
add linear combination instruction.

we can optimize this linear combination operation on plonk gate, specialized at ARITY

tsunrise committed:
plonk optimization

this optimization optimized ARITY-3 poseidon from 1124 to 433 gates

Commit:
profile native

tsunrise committed:
benchmarks (plonk prove & verify)

there is a bug during `verify_proof` though

tsunrise committed:
fix in poseidon benchmarks

tsunrise committed:
adapt grain lfsr from ark-sponge, but we have a bug here

tsunrise committed:
lfsr consistent

tsunrise committed:
parameter matches sage script

tsunrise committed:
minor

Commit:
poseidon r1cs profiling

tsunrise committed:
add hardcoded test for poseidon_ref

tsunrise committed:
remove neptune dependency because we have hardcoded test now

tsunrise committed:
nit

tsunrise committed:
impl hash trait, remove dead code

tsunrise committed:
impl trait useful for merkle tree

Commit:
fix issue in bench; add comments

tsunrise committed:
update mds generation sage script

tsunrise commented:
moved to https://github.com/Manta-Network/manta-rs/pull/57
